service/src/main/java/bio/terra/catalog/service/DatasetService.java

```diff
@@ -131,4 +144,37 @@ public DatasetId createDataset(
     ensureActionPermission(user, dataset, SamAction.CREATE_METADATA);
     return datasetDao.create(dataset).id();
   }
+
+  public DatasetPreviewTablesResponse getDatasetPreviewTables(
+      AuthenticatedUserRequest user, DatasetId datasetId) {
+    var dataset = datasetDao.retrieve(datasetId);
+    ensureActionPermission(user, dataset, SamAction.READ_ANY_METADATA);
```

Comment:
Looking at this I'm realizing that the code in `DatarepoService.userHasAction()` isn't quite correct; it combines reader and discoverer operations and does the same check for both. So this check will pass even if a user only has discoverer permissions. Since TDR itself also does a permission check, and uses the current user's credentials, no data will be returned and the TDR API call will return an error (probably 403).

For now this is OK, but this makes me wonder if we should be doing a permission check here at all. For operations that only pass through to other systems, we're duplicating work done elsewhere and therefore have a chance to do it incorrectly (as we're doing now). For operations that modify the catalog database, this check is still necessary.

Reply:
I created [DC-315](https://broadworkbench.atlassian.net/browse/DC-315) to track this issue.